cmd/contour/serve.go

```diff
@@ -380,46 +383,48 @@ func (s *Server) doServe() error {
 	}
 
-	// Build the core Kubernetes event handler.
-	contourHandler := &contour.EventHandler{
-		HoldoffDelay:    100 * time.Millisecond,
-		HoldoffMaxDelay: 500 * time.Millisecond,
-		Observer:        dag.ComposeObservers(append(xdscache.ObserversOf(resources), snapshotHandler)...),
-		Builder: s.getDAGBuilder(dagBuilderConfig{
-			ingressClassName:          ingressClassName,
-			rootNamespaces:            contourConfiguration.HTTPProxy.RootNamespaces,
-			gatewayAPIConfigured:      contourConfiguration.Gateway != nil,
-			disablePermitInsecure:     contourConfiguration.HTTPProxy.DisablePermitInsecure,
-			enableExternalNameService: contourConfiguration.EnableExternalNameService,
-			dnsLookupFamily:           contourConfiguration.Envoy.Cluster.DNSLookupFamily,
-			headersPolicy:             contourConfiguration.Policy,
-			clientCert:                clientCert,
-			fallbackCert:              fallbackCert,
-		}),
-		FieldLogger: s.log.WithField("context", "contourEventHandler"),
+	sh := &k8s.StatusUpdateHandler{
+		Log:    s.log.WithField("context", "StatusUpdateHandler"),
+		Client: s.mgr.GetClient(),
 	}
+	if err := s.mgr.Add(sh); err != nil {
+		return err
+	}
+
+	builder := s.getDAGBuilder(dagBuilderConfig{
+		ingressClassName:          ingressClassName,
+		rootNamespaces:            contourConfiguration.HTTPProxy.RootNamespaces,
+		gatewayAPIConfigured:      contourConfiguration.Gateway != nil,
+		disablePermitInsecure:     contourConfiguration.HTTPProxy.DisablePermitInsecure,
+		enableExternalNameService: contourConfiguration.EnableExternalNameService,
+		dnsLookupFamily:           contourConfiguration.Envoy.Cluster.DNSLookupFamily,
+		headersPolicy:             contourConfiguration.Policy,
+		clientCert:                clientCert,
+		fallbackCert:              fallbackCert,
+	})
+
+	// Build the core Kubernetes event handler.
+	contourHandler := contour.NewEventHandler(
+		contour.EventHandlerConfig{
+			Logger:          s.log.WithField("context", "contourEventHandler"),
+			HoldoffDelay:    100 * time.Millisecond,
+			HoldoffMaxDelay: 500 * time.Millisecond,
+			Observer: &contour.RebuildMetricsObserver{
+				Metrics:      contourMetrics,
+				IsLeader:     s.mgr.Elected(),
+				NextObserver: dag.ComposeObservers(append(xdscache.ObserversOf(resources), snapshotHandler)...),
+			},
+			StatusUpdater: sh.Writer(),
+			Builder:       builder,
+		},
+	)
 
 	// Wrap contourHandler in an EventRecorder which tracks API server events.
 	eventHandler := &contour.EventRecorder{
 		Next:    contourHandler,
 		Counter: contourMetrics.EventHandlerOperations,
 	}
 
-	// Register leadership election.
-	if s.ctx.DisableLeaderElection {
-		contourHandler.IsLeader = disableLeaderElection(s.log)
-	} else {
-		contourHandler.IsLeader = setupLeadershipElection(&s.group, s.log, s.ctx.Config.LeaderElection, s.coreClient, contourHandler.UpdateNow)
```

Comment:
this code used to trigger a DAG rebuild immediately when a Contour instance was elected leader, which helped avoid stale status issues - I'm not seeing anything that will do that now, but could be missing it.

Reply:
yeah thats a good point, I'll try to think up a test that exercises that

Reply:
the e2e test i can think of is a little racy:
1. delete leader pod
2. make a change to an httpproxy/gateway/gatewayclass/etc. that would make it invalid
3. eventually resource should have correct status

however, 2. needs to be timed just right, after the original leader has stopped but before a new leader has been elected

Reply:
simple fix would be to replicate what we had before, and pass in the manager "elected" channel to the event handler, so it re-runs on leader election

longer-term, seems we might need to have coverage over this logic with a controller-runtime style test of the controller itself: https://github.com/projectcontour/contour/blob/7bf2309d411a11a80b1e0f75918f9ea5b8932d78/internal/controller/gateway.go#L89-L107 (and propagate it to other controllers, not just gateway and gatewayclass)

this way we can more easily control and test how these processes are coordinated

we might also want to possibly move other things to controllers rather than informer-based, e.g. httpproxy/ingress/secrets etc. to standardize testing

Reply:
One idea for the E2E, could you manually create the leader-elect configmap before starting up Contour, indicating some non-existent pod holds the lease? That way no Contour pod would acquire it for at least a few seconds. Still some raciness, but maybe would be easier than having to delete a pod. Food for thought.

Reply:
another race i think is that even if you do trigger a dag update/status update handler update on leader election, if that isn't running yet (bc some processes start later), you might miss a write to appropriate channels and miss an update

Reply:
Another idea for the implementation here would be to bring back the event handler's `UpdateNow` function, then define a new leader-elected Runnable that, when started, just calls `eventHandler.UpdateNow()`. Shouldn't be much different from what you have now, but maybe a little simpler?

Reply:
https://github.com/projectcontour/contour/pull/4202/commits/602f8f7c1d1258136d3d7c4ff2f686c7865513c0

wdyt, used that idea ^ rather than having channels flying around